File: packages/go/kreuzberg/metadata.go
package kreuzberg

import "encoding/json"

var metadataCoreKeys = map[string]struct{}{
	"language":            {},
	"date":                {},
	"subject":             {},
	"format_type":         {},
	"image_preprocessing": {},
	"json_schema":         {},
	"error":               {},
}

var formatFieldSets = map[string][]string{
	"pdf": {
		"title", "subject", "authors", "keywords", "created_at", "modified_at",
		"created_by", "producer", "page_count", "pdf_version", "is_encrypted",
		"width", "height", "summary",
	},
	"excel":   {"sheet_count", "sheet_names"},
	"email":   {"from_email", "from_name", "to_emails", "cc_emails", "bcc_emails", "message_id", "attachments"},
	"pptx":    {"title", "author", "description", "summary", "fonts"},
	"archive": {"format", "file_count", "file_list", "total_size", "compressed_size"},
	"image":   {"width", "height", "format", "exif"},
	"xml":     {"element_count", "unique_elements"},
	"text":    {"line_count", "word_count", "character_count", "headers", "links", "code_blocks"},
	"html": {
		"title", "description", "keywords", "author", "canonical", "base_href",
		"og_title", "og_description", "og_image", "og_url", "og_type", "og_site_name",
		"twitter_card", "twitter_title", "twitter_description", "twitter_image", "twitter_site", "twitter_creator",
		"link_author", "link_license", "link_alternate",
	},
	"ocr": {"language", "psm", "output_format", "table_count", "table_rows", "table_cols"},
}

// UnmarshalJSON ensures Metadata captures flattened format unions and additional custom fields.
func (m *Metadata) UnmarshalJSON(data []byte) error {
	type alias Metadata
	var aux struct {
		Language           *string                     `json:"language"`
		Date               *string                     `json:"date"`
		Subject            *string                     `json:"subject"`
		FormatType         string                      `json:"format_type"`
		ImagePreprocessing *ImagePreprocessingMetadata `json:"image_preprocessing"`
		JSONSchema         json.RawMessage             `json:"json_schema"`
		Error              *ErrorMetadata              `json:"error"`
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	m.Language = aux.Language
	m.Date = aux.Date
	m.Subject = aux.Subject
	m.FormatType = aux.FormatType
	m.ImagePreprocessing = aux.ImagePreprocessing
	m.JSONSchema = aux.JSONSchema
	m.Error = aux.Error

	if err := m.decodeFormat(data); err != nil {
		return err
	}

	raw := map[string]json.RawMessage{}
	if err := json.Unmarshal(data, &raw); err == nil {
		m.Additional = make(map[string]json.RawMessage)
		for key, value := range raw {
			if _, ok := metadataCoreKeys[key]; ok {
				continue
			}
			if m.isFormatField(key) {
				continue
			}
			m.Additional[key] = value
		}
		if len(m.Additional) == 0 {
			m.Additional = nil
		}
	}

	return nil
}

func (m *Metadata) decodeFormat(data []byte) error {
	switch m.FormatType {
	case "pdf":
		var meta PdfMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Pdf = &meta
		} else {
			return err
		}
	case "excel":
		var meta ExcelMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Excel = &meta
		} else {
			return err
		}
	case "email":
		var meta EmailMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Email = &meta
		} else {
			return err
		}
	case "pptx":
		var meta PptxMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Pptx = &meta
		} else {
			return err
		}
	case "archive":
		var meta ArchiveMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Archive = &meta
		} else {
			return err
		}
	case "image":
		var meta ImageMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Image = &meta
		} else {
			return err
		}
	case "xml":
		var meta XMLMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.XML = &meta
		} else {
			return err
		}
	case "text":
		var meta TextMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.Text = &meta
		} else {
			return err
		}
	case "html":
		var meta HtmlMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.HTML = &meta
		} else {
			return err
		}
	case "ocr":
		var meta OcrMetadata
		if err := json.Unmarshal(data, &meta); err == nil {
			m.OCR = &meta
		} else {
			return err
		}
	}
	return nil
}

func (m *Metadata) isFormatField(key string) bool {
	if m.FormatType == "" {
		return false
	}
	if fields, ok := formatFieldSets[m.FormatType]; ok {
		for _, candidate := range fields {
			if candidate == key {
				return true
			}
		}
	}
	return false
}

File: packages/typescript/core/src/types/metadata.ts
/**
 * Metadata type definitions for extracted documents.
 *
 * Includes format-specific metadata for various document types
 * (PDF, Excel, Email, Images, etc.) and the unified Metadata interface.
 */

// ============================================================================
// Format-Specific Metadata Interfaces
// ============================================================================

export interface ExcelMetadata {
	sheetCount?: number;
	sheetNames?: string[];
}

export interface EmailMetadata {
	fromEmail?: string | null;
	fromName?: string | null;
	toEmails?: string[];
	ccEmails?: string[];
	bccEmails?: string[];
	messageId?: string | null;
	attachments?: string[];
}

export interface ArchiveMetadata {
	format?: string;
	fileCount?: number;
	fileList?: string[];
	totalSize?: number;
	compressedSize?: number | null;
}

export interface ImageMetadata {
	width?: number;
	height?: number;
	format?: string;
	exif?: Record<string, string>;
}

export interface XmlMetadata {
	elementCount?: number;
	uniqueElements?: string[];
}

export interface TextMetadata {
	lineCount?: number;
	wordCount?: number;
	characterCount?: number;
	headers?: string[] | null;
	links?: [string, string][] | null;
	codeBlocks?: [string, string][] | null;
}

export interface HeaderMetadata {
	level: number;
	text: string;
	id?: string | null;
	depth: number;
	htmlOffset: number;
}

export interface LinkMetadata {
	href: string;
	text: string;
	title?: string | null;
	linkType: "anchor" | "internal" | "external" | "email" | "phone" | "other";
	rel: string[];
	attributes: Record<string, string>;
}

export interface ImageMetadata {
	src: string;
	alt?: string | null;
	title?: string | null;
	dimensions?: [number, number] | null;
	imageType: "data_uri" | "inline_svg" | "external" | "relative";
	attributes: Record<string, string>;
}

export interface StructuredData {
	dataType: "json_ld" | "microdata" | "rdfa";
	rawJson: string;
	schemaType?: string | null;
}

export interface HtmlMetadata {
	title?: string | null;
	description?: string | null;
	keywords: string[];
	author?: string | null;
	canonicalUrl?: string | null;
	baseHref?: string | null;
	language?: string | null;
	textDirection?: "ltr" | "rtl" | "auto" | null;
	openGraph: Record<string, string>;
	twitterCard: Record<string, string>;
	metaTags: Record<string, string>;
	htmlHeaders: HeaderMetadata[];
	htmlLinks: LinkMetadata[];
	htmlImages: ImageMetadata[];
	structuredData: StructuredData[];
}

export interface PdfMetadata {
	title?: string | null;
	author?: string | null;
	subject?: string | null;
	keywords?: string | null;
	creator?: string | null;
	producer?: string | null;
	creationDate?: string | null;
	modificationDate?: string | null;
	pageCount?: number;
}

export interface PptxMetadata {
	title?: string | null;
	author?: string | null;
	description?: string | null;
	summary?: string | null;
	fonts?: string[];
}

export interface OcrMetadata {
	language?: string;
	psm?: number;
	outputFormat?: string;
	tableCount?: number;
	tableRows?: number | null;
	tableCols?: number | null;
}

// ============================================================================
// Image Preprocessing Metadata
// ============================================================================

export interface ImagePreprocessingMetadata {
	originalDimensions?: [number, number];
	originalDpi?: [number, number];
	targetDpi?: number;
	scaleFactor?: number;
	autoAdjusted?: boolean;
	finalDpi?: number;
	newDimensions?: [number, number] | null;
	resampleMethod?: string;
	dimensionClamped?: boolean;
	calculatedDpi?: number | null;
	skippedResize?: boolean;
	resizeError?: string | null;
}

// ============================================================================
// Error Metadata
// ============================================================================

export interface ErrorMetadata {
	errorType?: string;
	message?: string;
}

// ============================================================================
// Unified Metadata Interface
// ============================================================================

/**
 * Extraction result metadata.
 *
 * Uses a flattened discriminated union approach with format_type as the discriminator.
 * When format_type is set (e.g., "archive"), the corresponding format-specific fields
 * are available at the root level of the metadata object.
 *
 * This structure matches the Rust serialization with serde's tagged enum flattening.
 */
export interface Metadata {
	language?: string | null;
	date?: string | null;
	subject?: string | null;

	format_type?: "pdf" | "excel" | "email" | "pptx" | "archive" | "image" | "xml" | "text" | "html" | "ocr";

	// Common PDF/Document metadata
	title?: string | null;
	author?: string | null;
	keywords?: string | null;
	creator?: string | null;
	producer?: string | null;
	creation_date?: string | null;
	modification_date?: string | null;
	page_count?: number;

	// Excel-specific metadata
	sheet_count?: number;
	sheet_names?: string[];

	// Email-specific metadata
	from_email?: string | null;
	from_name?: string | null;
	to_emails?: string[];
	cc_emails?: string[];
	bcc_emails?: string[];
	message_id?: string | null;
	attachments?: string[];

	// PowerPoint-specific metadata
	description?: string | null;
	summary?: string | null;
	fonts?: string[];

	// Archive-specific metadata
	format?: string;
	file_count?: number;
	file_list?: string[];
	total_size?: number;
	compressed_size?: number | null;

	// Image-specific metadata
	width?: number;
	height?: number;
	exif?: Record<string, string>;

	// XML-specific metadata
	element_count?: number;
	unique_elements?: string[];

	// Text-specific metadata
	line_count?: number;
	word_count?: number;
	character_count?: number;
	headers?: string[] | null;
	links?: [string, string][] | null;
	code_blocks?: [string, string][] | null;

	// HTML-specific metadata
	canonical_url?: string | null;
	base_href?: string | null;
	open_graph?: Record<string, string>;
	twitter_card?: Record<string, string>;
	meta_tags?: Record<string, string>;
	html_language?: string | null;
	text_direction?: "ltr" | "rtl" | "auto" | null;
	html_headers?: HeaderMetadata[];
	html_links?: LinkMetadata[];
	html_images?: ImageMetadata[];
	structured_data?: StructuredData[];

	// OCR-specific metadata
	psm?: number;
	output_format?: string;
	table_count?: number;
	table_rows?: number | null;
	table_cols?: number | null;

	// Image preprocessing metadata
	image_preprocessing?: ImagePreprocessingMetadata | null;

	// JSON schema
	json_schema?: Record<string, unknown> | null;

	// Error information
	error?: ErrorMetadata | null;

	// biome-ignore lint/suspicious/noExplicitAny: Postprocessors can add arbitrary metadata fields
	[key: string]: any;
}
